### ISSUE TITLE
Lookahead in forward 

### ISSUE THREAD

**Issue by Dionysour** [bitsandbytes-foundation/bitsandbytes#1447]
### System Info

transformers '4.47.0', bitsanbytes '0.45.0', torch '2.5.1+cu124', NVIDIA RTX A4000

### Reproduction

During assisted greedy generation with the verifier loaded in 8-bit, I have noticed that it diverges from vanilla greedy generation. I have investigated this problem and discovered lookahead during forward on a particular sequence. Here's a snippet of the notebook without unnecessary code. You can run it and see the problem, but you need a Llama2 chat token because it's the model I was using.

https://gist.github.com/Dionysour/24b352bb685f7d4a8ffd18896455700d

### Expected behavior

change of [n] token doesn't affect logits of [:n] tokens

**Comment by matthewdouglas** [comment]
Thanks for the report. I was unable to reproduce this with Llama 3.1 8B in an environment with torch 2.5.1+cu124, transformers 4.47.0, bitsandbytes 0.45.0, on an RTX 4090.

I'm downloading Llama2 now and will give that a try.

**Comment by Dionysour** [comment]
I think this behaviour with llama 3.1 is expected. The bug is very model and sequence dependent. I saw in on different sequences for different models. So to reproduce it model must be "[meta-llama/Llama-2-7b-chat-hf](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf)"

**Comment by poedator** [comment]
I also have a reproduction of the issue found by @Dionysour  in this notebook.
https://nbviewer.org/gist/poedator/52440da0cdab808a4c0ff929f009ae31

tested on A100 with
bitsandbytes 0.44.1
torch 2.4.1
transformers 4.45.2
python 3.9

Apparently something is wrong with how bnb-8bit model deals with attention.

**Comment by matthewdouglas** [comment]
Hi all,
Since this behavior is reproducible in older versions of bitsandbytes, I'm going to mark it as a lower priority. We'll welcome any PRs or contributions. If there is sufficient reasoning why this needs to be a higher priority please let us know.

**Comment by TimDettmers** [comment]
It seems this issue was happening in an old version of bitsandbytes. If anyone encounters this with an newer version of bitsandbytes, please reopen this issue. Thank you!
